File: mac/VibeTunnel/Core/Models/GitRepository.swift
import Foundation

/// Represents the current state and metadata of a Git repository.
///
/// `GitRepository` provides a comprehensive snapshot of a Git repository's status,
/// including file change counts, current branch, and remote URL information.
/// It's designed to be used with ``GitRepositoryMonitor`` for real-time monitoring
/// of repository states in the VibeTunnel menu bar interface.
public struct GitRepository: Sendable, Equatable, Hashable {
    // MARK: - Properties

    /// The root path of the Git repository (.git directory's parent)
    public let path: String

    /// Number of modified files
    public let modifiedCount: Int

    /// Number of added files
    public let addedCount: Int

    /// Number of deleted files
    public let deletedCount: Int

    /// Number of untracked files
    public let untrackedCount: Int

    /// Current branch name
    public let currentBranch: String?

    /// GitHub URL for the repository (cached, not computed)
    public let githubURL: URL?

    // MARK: - Computed Properties

    /// Whether the repository has uncommitted changes
    public var hasChanges: Bool {
        modifiedCount > 0 || addedCount > 0 || deletedCount > 0 || untrackedCount > 0
    }

    /// Total number of files with changes
    public var totalChangedFiles: Int {
        modifiedCount + addedCount + deletedCount + untrackedCount
    }

    /// Folder name for display
    public var folderName: String {
        URL(fileURLWithPath: path).lastPathComponent
    }

    /// Status text for display
    public var statusText: String {
        if !hasChanges {
            return "clean"
        }

        var parts: [String] = []
        if modifiedCount > 0 {
            parts.append("\(modifiedCount)M")
        }
        if addedCount > 0 {
            parts.append("\(addedCount)A")
        }
        if deletedCount > 0 {
            parts.append("\(deletedCount)D")
        }
        if untrackedCount > 0 {
            parts.append("\(untrackedCount)U")
        }
        return parts.joined(separator: " ")
    }

    // MARK: - Lifecycle

    public init(
        path: String,
        modifiedCount: Int = 0,
        addedCount: Int = 0,
        deletedCount: Int = 0,
        untrackedCount: Int = 0,
        currentBranch: String? = nil,
        githubURL: URL? = nil
    ) {
        self.path = path
        self.modifiedCount = modifiedCount
        self.addedCount = addedCount
        self.deletedCount = deletedCount
        self.untrackedCount = untrackedCount
        self.currentBranch = currentBranch
        self.githubURL = githubURL
    }

    // MARK: - Internal Methods

    /// Extract GitHub URL from a repository path
    static func getGitHubURL(for repoPath: String) -> URL? {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process.arguments = ["remote", "get-url", "origin"]
        process.currentDirectoryURL = URL(fileURLWithPath: repoPath)

        let outputPipe = Pipe()
        process.standardOutput = outputPipe
        process.standardError = Pipe()

        do {
            try process.run()
            process.waitUntilExit()

            guard process.terminationStatus == 0 else {
                return nil
            }

            let outputData = outputPipe.fileHandleForReading.readDataToEndOfFile()
            let output = String(data: outputData, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

            return parseGitHubURL(from: output)
        } catch {
            return nil
        }
    }

    /// Parse GitHub URL from git remote output
    static func parseGitHubURL(from remoteURL: String) -> URL? {
        // Handle HTTPS URLs: https://github.com/user/repo.git
        if remoteURL.hasPrefix("https://github.com/") {
            let cleanURL = remoteURL.hasSuffix(".git") ? String(remoteURL.dropLast(4)) : remoteURL
            return URL(string: cleanURL)
        }

        // Handle SSH URLs: git@github.com:user/repo.git
        if remoteURL.hasPrefix("git@github.com:") {
            let pathPart = String(remoteURL.dropFirst("git@github.com:".count))
            let cleanPath = pathPart.hasSuffix(".git") ? String(pathPart.dropLast(4)) : pathPart
            return URL(string: "https://github.com/\(cleanPath)")
        }

        return nil
    }
}

File: tauri/src-tauri/src/git_repository.rs
use serde::{Deserialize, Serialize};
use std::path::Path;
use std::process::Command;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct GitRepository {
    pub path: String,
    pub modified_count: usize,
    pub added_count: usize,
    pub deleted_count: usize,
    pub untracked_count: usize,
    pub current_branch: Option<String>,
    pub github_url: Option<String>,
}

impl GitRepository {
    pub fn new(path: String) -> Self {
        Self {
            path,
            modified_count: 0,
            added_count: 0,
            deleted_count: 0,
            untracked_count: 0,
            current_branch: None,
            github_url: None,
        }
    }

    pub fn has_changes(&self) -> bool {
        self.modified_count > 0
            || self.added_count > 0
            || self.deleted_count > 0
            || self.untracked_count > 0
    }

    pub fn total_changed_files(&self) -> usize {
        self.modified_count + self.added_count + self.deleted_count + self.untracked_count
    }

    pub fn folder_name(&self) -> &str {
        Path::new(&self.path)
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
    }

    pub fn status_text(&self) -> String {
        if !self.has_changes() {
            return "clean".to_string();
        }

        let mut parts = Vec::new();
        if self.modified_count > 0 {
            parts.push(format!("{}M", self.modified_count));
        }
        if self.added_count > 0 {
            parts.push(format!("{}A", self.added_count));
        }
        if self.deleted_count > 0 {
            parts.push(format!("{}D", self.deleted_count));
        }
        if self.untracked_count > 0 {
            parts.push(format!("{}U", self.untracked_count));
        }
        parts.join(" ")
    }

    /// Extract GitHub URL from a repository path
    pub fn get_github_url(repo_path: &str) -> Option<String> {
        let output = Command::new("git")
            .args(&["remote", "get-url", "origin"])
            .current_dir(repo_path)
            .output()
            .ok()?;

        if !output.status.success() {
            return None;
        }

        let remote_url = String::from_utf8(output.stdout)
            .ok()?
            .trim()
            .to_string();

        Self::parse_github_url(&remote_url)
    }

    /// Parse GitHub URL from git remote output
    fn parse_github_url(remote_url: &str) -> Option<String> {
        // Handle HTTPS URLs: https://github.com/user/repo.git
        if remote_url.starts_with("https://github.com/") {
            let clean_url = if remote_url.ends_with(".git") {
                &remote_url[..remote_url.len() - 4]
            } else {
                remote_url
            };
            return Some(clean_url.to_string());
        }

        // Handle SSH URLs: git@github.com:user/repo.git
        if remote_url.starts_with("git@github.com:") {
            let path_part = &remote_url["git@github.com:".len()..];
            let clean_path = if path_part.ends_with(".git") {
                &path_part[..path_part.len() - 4]
            } else {
                path_part
            };
            return Some(format!("https://github.com/{}", clean_path));
        }

        None
    }
}